src/osd/scrubber/scrub_machine.cc

```diff
@@ -789,75 +793,109 @@ ReplicaActive::~ReplicaActive()
-    next_action = ReplicaReactCode::goto_waiting_reservation;
+    reserver.request_reservation(pg_id, reservation_cb, /*prio=*/0, nullptr);
+    m_reservation_status = reservation_status_t::requested_or_granted;
 
   } else {
     // an immediate yes/no is required
     Message* reply{nullptr};
     reservation_granted = reserver.request_reservation_or_fail(pg_id);
     if (reservation_granted) {
       dout(10) << fmt::format("{}: reserved? yes", __func__) << dendl;
+      m_reservation_status = reservation_status_t::requested_or_granted;
       reply = new MOSDScrubReserve(
 	  spg_t(pg_id.pgid, m_pg->get_primary().shard), ev.m_op->sent_epoch,
-	  MOSDScrubReserve::GRANT, m_pg->pg_whoami, m->reservation_nonce);
-      next_action = ReplicaReactCode::goto_replica_reserved;
+	  MOSDScrubReserve::GRANT, m_pg->pg_whoami, m.reservation_nonce);
 
     } else {
       dout(10) << fmt::format("{}: reserved? no", __func__) << dendl;
       reply = new MOSDScrubReserve(
 	  spg_t(pg_id.pgid, m_pg->get_primary().shard), ev.m_op->sent_epoch,
-	  MOSDScrubReserve::REJECT, m_pg->pg_whoami, m->reservation_nonce);
-      // the event is discarded
-      next_action = ReplicaReactCode::discard;
+	  MOSDScrubReserve::REJECT, m_pg->pg_whoami, m.reservation_nonce);
     }
 
     m_osds->send_message_osd_cluster(
 	reply, ev.m_op->get_req()->get_connection());
   }
-
-  return next_action;
+  return discard_event();
 }
 
-bool ReplicaActive::granted_by_reserver(const AsyncScrubResData& reservation)
+
+sc::result ReplicaActive::react(const ReserverGranted& ev)
 {
   DECLARE_LOCALS;  // 'scrbr' & 'pg_id' aliases
-  dout(10) << fmt::format("{}: reservation granted: {}", __func__, reservation)
-	   << dendl;
+  const AsyncScrubResData& reservation = ev.value;
+  dout(10)
+      << fmt::format(
+	     "ReplicaActive::react(const ReserverGranted&). Reservation:{}",
+	     reservation)
+      << dendl;
+
+  /**
+   * discard (and log) unexpected 'reservation granted' messages
+   * from the async reserver.
+   * As canceled reservations may still be triggered, this is not
+   * necessarily a bug.
+   */
+  if (!pending_reservation_nonce) {
+    dout(5) << fmt::format(
+		   "ReplicaActive::react(const ReserverGranted&): reservation "
+		   "granted while not being waited for{} (pending_nonce:{})",
+		   __func__, reservation, pending_reservation_nonce)
+	    << dendl;
+    return discard_event();
+  }
 
   /// verify that the granted reservation is the one we were waiting for
   if (reservation.nonce != pending_reservation_nonce) {
     dout(5) << fmt::format(
-	"{}: reservation_nonce mismatch: {} != {}", __func__, reservation.nonce,
-	pending_reservation_nonce) << dendl;
-    return false;
+		   "{}: reservation_nonce mismatch: {} != {}", __func__,
+		   reservation.nonce, pending_reservation_nonce)
+	    << dendl;
+    // not investing the effort to handle this gracefully
+    return transit<ReplicaActive>();
```

Comment:
I think this is actually possible (which is why we record a pending_reservation_nonce) and the correct behavior is to ignore it (```discard_event()```) rather than to transit to ReplicaActive.  If the replica cancels a pending reservation before it is granted, there's a possible ordering like this:
1. reserver submits ReserverGranted event
2. replica cancels the reservation due to a message from the primary and resets by transiting to ReplicaActive
3. replica starts a new reservation due to a message from the primary
4. replica processes message from 1.

Reply:
Same for the !pending_reservation_nonce case above -- you can handle them with a single comparison rather than two separate conditions.